app/models/relation.rb

```diff
@@ -248,6 +248,9 @@ def preconditions_ok?(good_members = [])
       model = Kernel.const_get(m[0].capitalize)
       # get the element with that ID
       element = model.find_by(:id => m[1])
+      # if found, lock the element to ensure it can't be deleted until
+      # after the current transaction commits.
+      element.lock!("for share") unless element.nil?
```

Comment:
The rails docs seem to suggest that `lock!` reloads the row, so wouldn't cause a race condition. However, this means its doing two queries where one would be enough, so it's definitely worth making your change.